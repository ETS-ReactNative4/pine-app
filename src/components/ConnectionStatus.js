import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';
import StyledText from '../components/StyledText';

const COLOR_ERROR = '#FF3B30';
const COLOR_WARNING = '#FF8D36';
const COLOR_SUCCESS = '#4CD964';

const LABEL_NOT_CONNECTED = 'Not Connected';
const LABEL_CONNECTED = 'Connected';
const LABEL_INCONSISTENT_NETWORKS = 'Inconsistent Networks';
const LABEL_SERVER_NOT_CONNECTED_TO_NODE = 'Server Not Connected to Node';

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  dot: {
    width: 9,
    height: 9,
    borderRadius: 5,
    marginRight: 12
  },
  label: {
    fontSize: 15,
    fontWeight: '600'
  }
});

export default class ConnectionStatus extends Component {
  render() {
    const { settings, serverInfo } = this.props;
    let dotColor = COLOR_ERROR;
    let status = LABEL_NOT_CONNECTED;

    if (serverInfo) {
      if (serverInfo.isConnected) {
        if (serverInfo.network === settings.bitcoin.network) {
          dotColor = COLOR_SUCCESS;
          status = LABEL_CONNECTED;
        } else {
          dotColor = COLOR_WARNING;
          status = LABEL_INCONSISTENT_NETWORKS;
        }
      } else if (serverInfo.isConnected === false) {
        status = LABEL_SERVER_NOT_CONNECTED_TO_NODE;
      }
    }

    return (
      <View style={styles.wrapper}>
        <View style={[styles.dot, { backgroundColor: dotColor }]} />
        <StyledText style={styles.label}>{status}</StyledText>
      </View>
    );
  }
}

ConnectionStatus.propTypes = {
  settings: PropTypes.object,
  serverInfo: PropTypes.object
};
