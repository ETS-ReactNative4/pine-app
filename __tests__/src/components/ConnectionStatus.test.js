import React from 'react';
import renderer from 'react-test-renderer';
import ConnectionStatus from '../../../src/components/ConnectionStatus';

describe('ConnectionStatus', () => {
  let fakeSettings;
  let fakeServerInfo;

  beforeEach(() => {
    fakeSettings = {
      bitcoin: {
        network: 'testnet'
      }
    };

    fakeServerInfo = {
      isConnected: true,
      network: 'testnet',
      blocks: 0
    };
  });

  it('renders Connected correctly', () => {
    const tree = renderer.create(
      <ConnectionStatus settings={fakeSettings} serverInfo={fakeServerInfo} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders Not Connected correctly', () => {
    fakeServerInfo = null;

    const tree = renderer.create(
      <ConnectionStatus settings={fakeSettings} serverInfo={fakeServerInfo} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders Inconsistent Networks correctly', () => {
    fakeServerInfo.network = 'mainnet';

    const tree = renderer.create(
      <ConnectionStatus settings={fakeSettings} serverInfo={fakeServerInfo} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders Server Not Connected to Node correctly', () => {
    fakeServerInfo.isConnected = false;

    const tree = renderer.create(
      <ConnectionStatus settings={fakeSettings} serverInfo={fakeServerInfo} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
