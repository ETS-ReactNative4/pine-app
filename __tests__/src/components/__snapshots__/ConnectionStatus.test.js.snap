// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionStatus renders Connected correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "height": 9,
          "marginRight": 12,
          "width": 9,
        },
        Object {
          "backgroundColor": "#4CD964",
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "fontSize": 15,
          "fontWeight": "600",
        },
      ]
    }
  >
    Connected
  </Text>
</View>
`;

exports[`ConnectionStatus renders Inconsistent Networks correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "height": 9,
          "marginRight": 12,
          "width": 9,
        },
        Object {
          "backgroundColor": "#FF8D36",
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "fontSize": 15,
          "fontWeight": "600",
        },
      ]
    }
  >
    Inconsistent Networks
  </Text>
</View>
`;

exports[`ConnectionStatus renders Not Connected correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "height": 9,
          "marginRight": 12,
          "width": 9,
        },
        Object {
          "backgroundColor": "#FF3B30",
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "fontSize": 15,
          "fontWeight": "600",
        },
      ]
    }
  >
    Not Connected
  </Text>
</View>
`;

exports[`ConnectionStatus renders Server Not Connected to Node correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "height": 9,
          "marginRight": 12,
          "width": 9,
        },
        Object {
          "backgroundColor": "#FF3B30",
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "fontSize": 15,
          "fontWeight": "600",
        },
      ]
    }
  >
    Server Not Connected to Node
  </Text>
</View>
`;
